### ISSUE TITLE
Inconsistent behaviour with Sides API

### ISSUE THREAD

**Issue by Column01** [MightyPirates/OpenComputers#3309]
## OpenComputers Versions:

-  `1.7.5.192` (core and main mod)

## Computer setup:

- Tier 3.5 memory
- Tier 3 case, screen, CPU, GPU 
- Tier 2 redstone card
- Internet Card
- Tier 1 hard disk drive
- Running OpenOS and LUA Bios

## Problem details
I was reacting to a redstone input on the `sides.right` side of the block (relative) and then based on some conditions writing to the `sides.back` of the block which again, relatively speaking was working fine.
Up to this point I was using `rs.getInput()` and `rs.setOutput()` to manipulate the redstone signals around the computer case and all was fine and dandy. Today I decided to change my code to use the `redstone_changed` event to reduce the time the code spent looping over and checking redstone values for no reason. When I made this change, I noticed that my signal on the `sides.right` side of the block is now showing a side value of `2.0` which in the Sides API corresponds to `sides.back`. Now at first glance, this doesn't seem that bad, but when you step back and realize what's going on it becomes annoying as all hell.

When I then react to an input on the `sides.north` side of the block, that correlates to the right side and I get the right input properly. But when I go to write the redstone output to the `sides.back` (aka the relative side of the block where the redstone event actually happens on the back of the block), it writes it to the back but the `redstone_changed` event recognizes that as a signal **on the north side of the block** which is the `sides.north` side which is also the `sides.back` therefore my output signal was being interpreted as my input.

I believe the sides API should really be changed to have `sides.back` etc. correspond to **relative** sides of the computer case block regardless and reserve `sides.north` and other cardinal directions to their literal cardinal directions. This would fix this inconsistent behaviour across the board and resolve many issues when coding for OpenComputers.

TL;DR The sides API is treating the redstone component read and write differently in respect to the directions the events occur on.

The below image is a picture of the setup. On the right, I have a wireless redstone device that pulses when I click a remote, and on the back is where my desired output would be. The (relative) right side of the block is pointing to the cardinal direction north and the back is to the west which when you use the redstone card and the `redstone_changed` event, causes the severe conflict.

![img](https://i.imgur.com/pp9Vl8y.png)

**Comment by Column01** [comment]
[Here](https://gist.github.com/ec76c5d60780ee080ca526f516666130) is my code that works (when the server isn't lagging at least lol). The second I change to the redstone event, it toggles itself basically lol. I'm using a custom event to toggle the redstone output of the back side of the block.

EDIT: For some reason that code is no longer working despite changing anything. Let me investigate
I change nothing on my original code and it no longer works.

EDIT 2: The sides API strikes again! I had it set to `sides.right` to read input and it was reading the back output as the input... This is a serious issue man. The code is fixed now although the directions used DO NOT correlate to in game cardinal directions LIKE THEY SHOULD MAN

**Comment by Column01** [comment]
TL;DR  number 2:
- `sides.back` when outputting is the back of the computer.
- `sides.back` when handling the input is the cardinal direction north which may or may not correspond to the proper back of the computer.

The real issue:
Inconsistent handling of relative sides of the computer. Either have `sides.back` ALWAYS be north or handle it as if it were the back of the computer (which makes 100 times more sense)

**Comment by payonel** [comment]
this was the same issue resolved for #3184
the internal redstone card should recieve, and communicate (i.e. get/set Inputs), all redstone signals as relative sides

you are experiencing a bug in our last release (1.7.5) that is fixed in our devs builds ( https://ci.cil.li/ )
you should be seeing sides.right for both the redstone_changes and getInput

Now...we also regret that our relative sides overlap our absolute sides. we should have separated those integer values in the beginning. but that's where we are are we're going to maintain that

**Comment by Column01** [comment]
> Now...we also regret that our relative sides overlap our absolute sides. we should have separated those integer values in the beginning. but that's where we are are we're going to maintain that

Is there a specific reasoning to keep it this way? Backwards compatibility I suppose but in reality, it just makes beginners wanna quit using it lol I almost did till I understood what was happening

**Comment by Column01** [comment]
If you don't plan on changing the sides API could you at least offer a way to get cardinal directions of the computer or the orientation in regard to the front of the block?

Like for example in the Computer API, add a method that is for example: `computer.getOrientation()` which returns `north: number, south: number, east: number, west: number` where each of those numbers corresponds to the actually block's cardinal directions and sides following the existing side's API numbering:
![img](https://i.imgur.com/7vg5gXg.png)

I.E. if the front of the computer was facing cardinal north, the first argument returned would be 3 for side three aka front.

**Comment by Column01** [comment]
I'm also curious what commit was this fixed in? Curious what went wrong :D

**Comment by payonel** [comment]
only the drone and robot can do that, with the navigation upgrade with gives them a `getFacing` api they can use. Not available to computers and servers.
i dont have the commit handy, but it was a regression i caused when expanding our bundled redstone support, i forgot to refactor the code to also make sure the side data was rotated relative to the case

**Comment by Column01** [comment]
Ah, I see that now that I read that linked issue more carefully.
As for the orientation of computers and servers, could we also allow the use of the navigation upgrade or a stripped-down version to allow us to get the block's real-world orientation? It would be very useful in certain bare-bones setups in my opinion.

**Comment by payonel** [comment]
this has come up before, and it is the benefit of the robot, to be able to move about the world ... we settled that navigation and orientation is a robot thing
